### PR TITLE
octeon: bring up target itusrouter, itusbridge, itusgateway

### PR DESCRIPTION
**Pull request by Grommish**
Itus Networks Shield - 1Ghz dual-core mips64 / Cavium Octeon 3 SoC, 1Gb RAM, 4Gb eMMC.

Signed-off-by: Donald Hoskins <grommish@gmail.com>

### CODE REVIEW

**Comment by Grommish** [comment]
Built on OpenWrt SNAPSHOT r13871+3-546e140382 / LuCI Master git-20.186.79919-0c47989, kernel 5.4
Tested on device

File: target/linux/octeon/base-files/lib/preinit/79_move_config [hunk 1]
```diff
@@ -10,6 +10,22 @@ move_config() {
 			[ -f "/mnt/$BACKUP_FILE" ] && mv -f "/mnt/$BACKUP_FILE" /
 			umount /mnt
 			;;
+		itusrouter)
```
Comment (dwmw2) [inline target/linux/octeon/base-files/lib/preinit/79_move_config:13]: Couldn't you have just one case here for `itusrouter|itusbridge|itusgateway` ?

Comment (Grommish) [reply target/linux/octeon/base-files/lib/preinit/79_move_config:13]: Yes, i was splitting them out for testing, but i don't know why I couldn't do a "itus"* there
Will correct

**Comment by adrianschmutzler** [comment]
> bring up target itusrouter, itusbridge, itusgateway

"target" -> "device"

That might also be the reason why I partially misunderstand you on IRC.

I'd however just use "octeon: add support for itusrouter, ..." anyway.

File: target/linux/octeon/base-files/etc/board.d/01_network [hunk 2]
```diff
@@ -8,6 +8,9 @@
 board_config_update
 
 case "$(board_name)" in
+itus*)
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/etc/board.d/01_network:11]: Please don't use wildcards:
https://openwrt.org/docs/guide-developer/device-support-policies#no_wildcards

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo [hunk 3]
```diff
@@ -1,31 +1,69 @@
...
+                ;;
+        esac
 
-	[ -e "/tmp/sysinfo/" ] || mkdir -p "/tmp/sysinfo/"
+        [ -e "/tmp/sysinfo/" ] || mkdir -p "/tmp/sysinfo/"
 
-	echo "$name" > /tmp/sysinfo/board_name
-	echo "$machine" > /tmp/sysinfo/model
+        echo "$name" > /tmp/sysinfo/board_name
+        echo "$machine" > /tmp/sysinfo/model
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/preinit/01_sysinfo:67]: Changing these indents is unrelated and shouldn't be done anyway, as we prefer tabs here (traditionally).

File: target/linux/octeon/base-files/etc/board.d/01_network (hunk 2 above)
Comment (Grommish) [reply target/linux/octeon/base-files/etc/board.d/01_network:11]: I need to use itus* or define itusrouter itusgateway itusbridge, and they all have the same default port configuration.

Which should I use?

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo [hunk 4]
```diff
@@ -1,31 +1,69 @@
...
+        else
+           if [ $GPIO17 -eq 0 ]; then
+              # Bridge Mode
+                 echo "itusbridge"
+           else
+              # Router Mode
+                 echo "itusrouter"
+           fi
+        fi
+}
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: Is there really no better way?

Actually, if there is no better, I'd actually prefer to make three images, and have the user choose the correct one instead of this detection mechanism.

File: target/linux/octeon/base-files/lib/preinit/79_move_config (hunk 1 above)
Comment (dwmw2) [reply target/linux/octeon/base-files/lib/preinit/79_move_config:13]: (from IRC discussion) ... because you actually want them to have different filenames like `/mnt/${BACKUP_FILE}-${board}` to prevent an upgrade intended for one partition being ingested by another partition if the mode switch is flipped at the wrong time.

Comment (adrianschmutzler) [reply target/linux/octeon/base-files/lib/preinit/79_move_config:13]: No itus* please.
Just
```
itusbridge|\
itusgateway|\
itusrouter)
```

File: target/linux/octeon/base-files/lib/upgrade/platform.sh [hunk 5]
```diff
@@ -28,6 +29,29 @@ platform_copy_config() {
...
 		umount /mnt
 		;;
+	itusrouter-test)
+		mkdir -p /rom
+		mount /dev/mmcblk1p2 /rom
+		PREINIT=1 mount_root
+		mount_root done
+		cp -af "$UPGRADE_BACKUP" "/rom/$BACKUP_FILE"
+		umount /rom
+			;;
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/upgrade/platform.sh:39]: Please remove your testing leftovers.

File: target/linux/octeon/base-files/lib/upgrade/platform.sh [hunk 6]
```diff
@@ -28,6 +29,29 @@ platform_copy_config() {
...
+		mount -t vfat /dev/mmcblk1p1 /mnt
+		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
+		umount /mnt
+			;;
+	itusbridge)
+		mount -t vfat /dev/mmcblk1p1 /mnt
+		cp -af "$UPGRADE_BACKUP" "/mnt/$BACKUP_FILE"
+		umount /mnt
+		;;
+	itusgateway)
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/upgrade/platform.sh:50]: Same as above, one block with three identifiers, please.

File: target/linux/octeon/base-files/lib/upgrade/platform.sh [hunk 7]
```diff
@@ -38,20 +62,34 @@ platform_do_flash() {
 	local rootfs=$4
 
 	mkdir -p /boot
-	mount -t vfat /dev/$kernel /boot
 
-	[ -f /boot/vmlinux.64 -a ! -L /boot/vmlinux.64 ] && {
+	if [[ $board == "itusrouter" || $board == "itusbridge" || $board == "itusgateway" ]]; then
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/upgrade/platform.sh:66]: No double brackets [[ please, that's bash syntax.

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo (hunk 4 above)
Comment (Grommish) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: This is in prep to moving these to their own Subtarget, which will not assign BOARD_NAME correctly, or that I couldn't.  This is how the device reads the multi-switch port to know what mode it's in.

File: target/linux/octeon/base-files/lib/upgrade/platform.sh [hunk 8]
```diff
@@ -68,22 +106,32 @@ platform_do_upgrade() {
...
 }
 
 platform_check_image() {
 	local board=$(board_name)
 
 	case "$board" in
 	er | \
-	erlite)
+	erlite | \
+	itus*)
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/upgrade/platform.sh:134]: No wildcards.

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo (hunk 4 above)
Comment (dwmw2) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: It could be inferred from the `root=` command line that u-boot provides, couldn't it? 

Or we could make u-boot put it somewhere in the device tree?

File: target/linux/octeon/image/Makefile [hunk 9]
```diff
@@ -47,4 +47,40 @@ define Device/ubnt_edgerouter-lite
 endef
 TARGET_DEVICES += ubnt_edgerouter-lite
 
+#
+# Itus Network Shield Profiles
+#
+ITUSROUTER_CMDLINE:=console=ttyS0,115200, root=/dev/mmcblk1p2 rootfstype=squashfs,ext4,f2fs rootwait
+define Device/itusrouter
```
Comment (adrianschmutzler) [inline target/linux/octeon/image/Makefile:54]: Please use the currently common vendor_model style here, e.g.
itus_shield-router

I'd also be happy if you could use the board name similar to what's done on modern DTS targets, i.e. itus,shield-router based on the suggested name.

File: target/linux/octeon/image/Makefile [hunk 10]
```diff
@@ -47,4 +47,40 @@ define Device/ubnt_edgerouter-lite
...
+  DEVICE_VENDOR := Itus Networks
+  DEVICE_MODEL := Shield Router
+  BOARD_NAME := itusrouter
```
Comment (adrianschmutzler) [inline target/linux/octeon/image/Makefile:57]: I'm not too familiar with non-DTS targets. Do we still need BOARD_NAME here?

File: target/linux/octeon/patches-5.4/130-itus_shield_support.patch [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+ 	     cvmx_board_type_to_string(octeon_bootinfo->board_type));
+ 
+--- a/arch/mips/pci/pci-octeon.c
++++ b/arch/mips/pci/pci-octeon.c
+@@ -211,7 +211,7 @@ const char *octeon_get_pci_interrupts(vo
+ 		return "AAABAAAAAAAAAAAAAAAAAAAAAAAAAAAA";
+ 	case CVMX_BOARD_TYPE_BBGW_REF:
+ 		return "AABCD";
+-	case CVMX_BOARD_TYPE_CUST_DSR1000N:
++	case CVMX_BOARD_TYPE_ITUS_SHIELD:
```
Comment (adrianschmutzler) [inline target/linux/octeon/patches-5.4/130-itus_shield_support.patch:39]: What's the point of renaming that constant?

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo (hunk 4 above)
Comment (Grommish) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: the Initramfs files don't use root=, which is how the images get loaded onto the device.  

This code is what sets /tmp/sysinfo/board_name, so if I can do it here and not have to do with elsewhere in the system..  Or, can someone suggest a better way?

File: target/linux/octeon/base-files/lib/upgrade/platform.sh [hunk 12]
```diff
@@ -68,22 +106,32 @@ platform_do_upgrade() {
...
+		;;
 	*)
 		return 1
 	esac
 
 	platform_do_flash $tar_file $board $kernel $rootfs
 
 	return 0
-	
+
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/upgrade/platform.sh:125]: That line can be removed at all.

File: target/linux/octeon/base-files/lib/preinit/79_move_config (hunk 1 above)
Comment (Grommish) [reply target/linux/octeon/base-files/lib/preinit/79_move_config:13]: > No itus* please.
> Just
> 
> ```
> itusbridge|\
> itusgateway|\
> itusrouter)
> ```

No problems.. Will be changed :)

**Comment by adrianschmutzler** [comment]
Please add a more extensive commit description with proper Specifications and Flashing instructions.

File: target/linux/octeon/base-files/etc/board.d/01_network (hunk 2 above)
Comment (adrianschmutzler) [reply target/linux/octeon/base-files/etc/board.d/01_network:11]: Read the reference. Just replace this by
```
itusbridge|\
itusgateway|\
itusrouter)
```

as already suggested elsewhere (or the more modern names after we have agreed on them).

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo (hunk 4 above)
Comment (adrianschmutzler) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: So, please, explain what's actually going on here? Is this one device with three different modes or is it three devices?

File: target/linux/octeon/base-files/etc/board.d/01_network (hunk 2 above)
Comment (Grommish) [reply target/linux/octeon/base-files/etc/board.d/01_network:11]: Will correct

File: target/linux/octeon/patches-5.4/130-itus_shield_support.patch (hunk 11 above)
Comment (Grommish) [reply target/linux/octeon/patches-5.4/130-itus_shield_support.patch:39]: When i did initial bring up, the device reported as CUST_DSR1000N as the board..  So, it was renamed so it would report ITUS_SHIELD correctly

Comment (adrianschmutzler) [reply target/linux/octeon/patches-5.4/130-itus_shield_support.patch:39]: but that's where I don't get it: you are changing the name of the variable here, not the content.

File: target/linux/octeon/image/Makefile (hunk 10 above)
Comment (Grommish) [reply target/linux/octeon/image/Makefile:57]: it's carried thru the build system as $(board_name) AFAIK, at least, which it isn't a $SUBTARGET, in which case I could only get it to put "itus" (as the subtarget) instead of the BOARD_NAME defined.

Comment (adrianschmutzler) [reply target/linux/octeon/image/Makefile:57]: subtarget is a completely different thing.

BOARD_NAME is mostly an old mechanism for SUPPORTED_DEVICES and also influences some file names in the image assembly process.

**Comment by mans0n** [comment]
I'm happy to see a new support PR for Octeon 3. :)
But isn't it supposed to be another subtarget? I thought we should add another line [here](https://github.com/openwrt/openwrt/blob/master/include/target.mk#L174). Does CPU_TYPE in image Makefile really work?

File: target/linux/octeon/patches-5.4/130-itus_shield_support.patch (hunk 11 above)
Comment (Grommish) [reply target/linux/octeon/patches-5.4/130-itus_shield_support.patch:39]: If I don't, I know that at least LUCI Shows it CUST_DSR1000N instead of ITUS_SHIELD (CN7020p1.2-1000-AAP) as the Model and Arch

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo (hunk 4 above)
Comment (Grommish) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: One device BOARD, but the switch determines which kernel image is chosen at boot, so THREE separate images (separate files, different root= defines, etc)

Comment (adrianschmutzler) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: So, at a reseller where you would buy it, there is actually only one "device" then?

Comment (adrianschmutzler) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo:34]: And the three images then just trigger different "configuration modes" for that some physical "device"?

**Comment by adrianschmutzler** [comment]
So, to me it appears that the only difference between your three "devices" will be the "root" parameter in cmdline and the partition chosen during upgrade?

This very much sounds to me like this should be one device in OpenWrt after all, and the distinction between the modes would only be used where necessary then.

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo (hunk 3 above)
Comment (Grommish) [reply target/linux/octeon/base-files/lib/preinit/01_sysinfo]: Will correct

File: target/linux/octeon/base-files/lib/upgrade/platform.sh (hunk 5 above)
Comment (Grommish) [reply target/linux/octeon/base-files/lib/upgrade/platform.sh]: Will correct

**Comment by Grommish** [comment]
> I'm happy to see a new support PR for Octeon 3. :)
> But isn't it supposed to be another subtarget? I thought we should add another line [here](https://github.com/openwrt/openwrt/blob/master/include/target.mk#L174). Does CPU_TYPE in image Makefile really work?

It did when I had split Itus out as a Subtarget separate from the others, but then I had it in the target.mk.  Should I split out Edgerouter and Itus into their own Subtargets?

**Comment by Grommish** [comment]
> So, to me it appears that the only difference between your three "devices" will be the "root" parameter in cmdline and the partition chosen during upgrade?
> 
> This very much sounds to me like this should be one device in OpenWrt after all, and the distinction between the modes would only be used where necessary then.

That would prevent automated builds if the device is included as a device, since any end-user would need to build their own image in order to set the root, no?

**Comment by adrianschmutzler** [comment]
> That would prevent automated builds if the device is included as a device, since any end-user would need to build their own image in order to set the root, no?

How is this switched with OEM firmware? Do they provide three different images as well? (still assuming that this is actually one physical device, as this question asked twice has not been answered so far)

**Comment by mans0n** [comment]
> Should I split out Edgerouter and Itus into their own Subtargets?

I would suggest `octeonplus` for edgerouter and octeon3 for itus.

> That would prevent automated builds if the device is included as a device, since any end-user would need to build their own image in order to set the root, no?

What is the stock firmware behavior? If you select one mode, then does it reboot to another root partition? Does the bootloader select a root partition depending on the switch mode?

**Comment by mans0n** [comment]
Come to think of it, if octeon+ and octeon3 share a same kernel config, then I would not split them into subtargets...

File: target/linux/octeon/patches-5.4/130-itus_shield_support.patch (hunk 11 above)
Comment (mans0n) [reply target/linux/octeon/patches-5.4/130-itus_shield_support.patch:39]: Probably the magic is in the `ENUM_BRD_TYPE_CASE`... I'm not sure if this is the best way to deal with it, but I can't think of any better idea.

File: target/linux/octeon/base-files/lib/upgrade/platform.sh [hunk 13]
```diff
@@ -38,20 +54,34 @@ platform_do_flash() {
 	local rootfs=$4
 
 	mkdir -p /boot
-	mount -t vfat /dev/$kernel /boot
 
-	[ -f /boot/vmlinux.64 -a ! -L /boot/vmlinux.64 ] && {
+	if [ $board == "itusrouter" || $board == "itusbridge" || $board == "itusgateway" ]; then
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/upgrade/platform.sh:58]: double == are bash, please use single =

File: target/linux/octeon/base-files/lib/preinit/01_sysinfo [hunk 14]
```diff
@@ -1,31 +1,69 @@
...
+	      ;;
 
+	"ITUS_SHIELD"*)
+	      name=$smode
+	      ;;
 	*)
-		name="generic"
-		;;
+	      name="generic"
+	      ;;
```
Comment (adrianschmutzler) [inline target/linux/octeon/base-files/lib/preinit/01_sysinfo:61]: Still indent issues here.

**Comment by Grommish** [comment]
It's been decided that the device would need to eliminate 2 of the 3 modes of operation and present only a single image.  Unfortunately, I don't think this will be a viable solution.  I'm going to go ahead and close the PR.
